---
title: "About_analysis"
output:
  workflowr::wflow_html:
    toc: false
editor_options:
  chunk_output_type: console
---
27 Nov 2024
Reworking this to make modular
`SetSimulationParameters.R`
`MakeTheFounderPopulation.R`

`MakeTheFounderPopulation.Rmd`
Sets the random seed
Sets a number of global parameters:
nChr  
nSnpPerChr  
nQTLperChr  
nFounders  
nProgeny  
Generates an object called `founders`  

`OptimizeWithinFamilyPrediction.Rmd` has all the same initial code as 
`MakeTheFounderPopulation.Rmd`  

## Taking the marker perspective to optimize the training population.  
Hypothesis: if an individual has a long stretch of IBS with a parent then it is
in fact IBD with the parent and it will better predict the parent effect in that
stretch.
1. Inbred individuals  
2. Each ind has a length IBS with each parent haplotype  
3. Maximize the sum of lengths  
4. Minimize the KS deviation from uniformity, done separately for each parent  
5. Minimize the difference in lengths between the two parents  

The basic idea is to construct a training population with these characteristics
by combinatorial optimization.  So pick a population and then iteratively 
improve it by swapping individuals and testing if a utility function increases.

Ultimately, the objective is to construct a training population in which the 
LD relationships between loci are as similar as possible to what they would be
among the progeny of the two parents. Fulfilling points 3 to 5 aims to achieve
that while not requiring that you calculate those LD relationships at each 
iteration of the optimization.

Noteworthy
1. The optimization can lead to TPs that have notably worse prediction
accuracy than a random population. I have yet to construct TPs that have notably
better accuracy. But just the fact of variation across the parameters of the
optimization is hopeful.
2. There is quite a bit of variation in the accuracy achieved among repeated
optimizations. I have not yet looked at whether characterization of the
constructed TP can illuminate what causes the differences in the ultimate
accuracy.

Tasks
1. See if very preliminary results are reproduced when I have higher SNP and QTL
densities. Hmmm: I can see that that is substantially slower.  Also, not any
better.  The optimization parameters might need to be revised.
2. Start setting up the approach for multiple chromosomes.
3. Start setting up the approach where the optimization is actually on the LD
relationships among the markers.
